core/modules/loader/CentralMaster.cc

```diff
@@ -0,0 +1,184 @@
+namespace qserv {
+namespace loader {
+
+
+void CentralMaster::addWorker(std::string const& ip, int udpPort, int tcpPort) {
+    auto item = _mWorkerList->addWorker(ip, udpPort, tcpPort);
+
+    if (item != nullptr) {
+        // If that was the first worker added, it gets unlimited range.
+        if (_firstWorkerRegistered.exchange(true) == false) {
+            LOGS(_log, LOG_LVL_INFO, "setAllInclusiveRange for name=" << item->getName());
+            item->setAllInclusiveRange();
+        }
+
+        item->addDoListItems(this);
+        LOGS(_log, LOG_LVL_INFO, "Master::addWorker " << *item);
+    }
+}
+
+
+void CentralMaster::updateWorkerInfo(uint32_t workerName, NeighborsInfo const& nInfo, StringRange const& strRange) {
+    if (workerName == 0) {
+        return;
+    }
+    auto item = getWorkerNamed(workerName);
+    if (item == nullptr) {
+        LOGS(_log, LOG_LVL_WARN, "CentralMaster::updateNeighbors nullptr for workerName=" << workerName);
+        return;
+    }
+    item->setNeighborsInfo(nInfo);
+    item->setRangeStr(strRange);
+    _assignNeighborIfNeeded();
+}
+
+
+void CentralMaster::setWorkerNeighbor(MWorkerListItem::WPtr const& target, int message, uint32_t neighborName) {
+    // Get the target worker's network address
+    auto targetWorker = target.lock();
+    if (targetWorker == nullptr) {
+        LOGS(_log, LOG_LVL_WARN, "CentralMaster::setWorkerNeighbor nullptr for" << neighborName);
+        return;
+    }
+
+    LOGS(_log, LOG_LVL_DEBUG, "CentralMaster::setWorkerNeighbor " << neighborName << " " << *targetWorker);
+    // Build and send the message
+    LoaderMsg msg(message, getNextMsgId(), getMasterHostName(), getMasterPort());
+    BufferUdp msgData;
+    msg.appendToData(msgData);
+    UInt32Element neighborNameElem(neighborName);
+    neighborNameElem.appendToData(msgData);
+    auto addr = targetWorker->getUdpAddress();
+    sendBufferTo(addr.ip, addr.port, msgData);
+}
+
+
+void CentralMaster::_assignNeighborIfNeeded() {
+    LOGS(_log, LOG_LVL_DEBUG, "CentralMaster::_assignNeighborIfNeeded");
+    // Go through the list and see if all the workers are full.
+    // If they are, assign a worker to the end (rightmost) worker
+    // and increase the maximum by an order of magnitude, max 10 million.
+    // TODO Make a better algorithm, insert workers at busiest worker.
+    // TODO maybe rate limit this check.
+    std::string funcName("_assignNeighborIfNeeded");
+    if (_addingWorker) {
+        // Already in process of adding a worker, so don't add another one.
+        // TODO Check if it failed.
+        return;
+    }
+    // Only one thread should ever be working on this logic at a time.
+    std::lock_guard<std::mutex> lck(_assignMtx);
+    auto pair = _mWorkerList->getActiveInactiveWorkerLists();
+    std::vector<MWorkerListItem::Ptr>& activeList = pair.first;
+    std::vector<MWorkerListItem::Ptr>& inactiveList = pair.second;
+    if (inactiveList.empty() || _addingWorker) { return; }
```

Comment:
The check on _addingWorker (which is atomic) earlier is meant to avoid trying to trying to lock _assignMtx, but as you said there is a race condition, so this second check is needed.